Add support for fields that require the modelId on creation - fixes #924

michaellindahl opened the pull request.

lloy0076 commented on src/PanelTraits/Create.php:
```diff
@@ -43,6 +51,24 @@ public function getCreateFields()
+        $fields = $this->create_fields;
+        $relationFields = [];
+
+        foreach ($fields as $field) {
```
I wonder if `array_filter` would be faster.

shaoshiva replied:
Or even better, use and return a collection :

```php
public function getAfterCreateFields()
{
    $relationFields = collect($this->create_fields)->filter(function($field) {
        return isset($field['after_create'];
    });

    return $relationFields;
}
```

Then replace :
```php
$after_create = array_pluck($this->getAfterCreateFields(), 'name');
```
... with :
```php
$after_create = $this->getAfterCreateFields()->pluck('name')->all();
```

lloy0076 commented on src/PanelTraits/Create.php:
```diff
@@ -21,14 +21,22 @@ public function create($data)
         $data = $this->compactFakeFields($data, 'create');
-
+        
         // ommit the n-n relationships when updating the eloquent item
```
// omit

lloy0076 commented on src/PanelTraits/Create.php:
```diff
@@ -21,14 +21,22 @@ public function create($data)
+        
+        $afterCreateData = array_only($data, $after_create);
+        foreach ($afterCreateData as $key => $value) {
+            $item->{$key} = $value;
```
Is it possible for `$value` to come up with `isset() !== true` and trigger an undefined key warning/error?

lloy0076 commented on src/PanelTraits/Create.php:
```diff
@@ -21,14 +21,22 @@ public function create($data)
+        $after_create = array_pluck($this->getAfterCreateFields(), 'name');
+        
+        $creationData = array_except($data, array_merge($nn_relationships, $after_create));
+        $item = $this->model->create($creationData);
```
That may fail - we probably should at minimum complain to the logs if it does.

lloy0076 commented on src/PanelTraits/Create.php:
```diff
@@ -21,14 +21,22 @@ public function create($data)
+        $afterCreateData = array_only($data, $after_create);
+        foreach ($afterCreateData as $key => $value) {
+            $item->{$key} = $value;
+        }
```
I forget if we'll need a `->save` on `$item`?

lloy0076 commented:
If this does get accepted it may warrant mention in the docs - I can't see any reason why it mightn't wend its way up (unless someone can think of any particular reason why it shouldn't).

tabacitu commented:
Hi @michaellindahl ,

If I understand correctly, this is for a 1-1 relationship, where the ID is stored on the primary table, not the secondary, right?

So:
- articles
   - id
   - name
   - image_id

- images
   - id
   - path

Something like that? It's a little unclear to me what the use case is, more exactly. Can you please clarify?

Thanks, cheers!

tabacitu commented:
Hmm... actually, after looking at the code again, it looks like it would work for 1-n relationships too. 

@michaellindahl could you please clarify how a developer would be able to fully use this? The way I understand it:
- you put ```after_create``` on your 1-n field;
- you modify ```EntityCrudController::store()``` to also add those connected entities?

Would love to merge it. I think this is a good feature to have, and want to explore the easiest way to achieve this, for the developer. But as-is, I don't understand how the connected entities would get created.

Thank you!

michaellindahl commented:
Sorry for the super late reply. I've been using this with a side project that I don't often touch. By day I'm an iOS developer.

I wanted this as I wrote my own simple image attachment models and I needed to attach a ImageModel to the CRUDModel before the crud model was completely created. I think I might move to using Spatie's Image Library for this eventually. They also do this to attach images to unsaved image models: https://twitter.com/freekmurze/status/973272542692683778

Right now I upgraded to the latest Backpack CRUD and without this fix it is working. I am closing this PR as I believe the approach that Spatie is doing is more appropriate if this indeed still needs fixing. But for now I'm not seeing any issues.

michaellindahl commented:
Update: So this issue wasn't fixed by updating the codebase. Forgot it only happens on creating new entities with images. However, I used the example from Laravel Media Library and I modified the existing trait that I was using to make this work. Here's an excerpt of my trait. 

Note that I did have to access the `$unAttachedImage` variable that I added with the trait via the provided methods on the trait else I ran into `Indirect modification of overloaded element` errors.


```
trait HasImage
{
    
    /** @var array */
    protected $unAttachedImage;

    /**
     * @return mixed
     */
    public function images()
    {
        return $this->morphMany(\App\Models\ModelImage::class, 'model_images', 'model_type', 'model_id');
    }

    public function setImageAttribute($value)
    {
        // delete the image from disk
        $this->images()->each(function ($item, $key) {
            \Storage::disk($item->disk)->delete($item->path);
            $item->delete();
        });

        if (! $this->exists) {
            $this->saveImageToAttachLater($value);

            $class = get_class($this);

            $class::created(function ($model) {
                if ($image = $this->imageToAttach()) {
                    dispatch(new CreateModelImages($image, $this));
                }
        
                $this->saveImageToAttachLater(null);
            });

            return;
        }
        
        dispatch(new CreateModelImages($value, $this));
    }

    public function saveImageToAttachLater($image)
    {
        $this->unAttachedImage = $image;
    }

    public function imageToAttach()
    {
        return $this->unAttachedImage;
    }
}
```

tabacitu commented:
Thank you for sharing your trait @michaellindahl ! 

Judging from your experience, and what a few others have been through too, I think we _should consider_ using spatie-media-library in 4.0 by default - that would fix this problem and simplify a few more things… Food for thought...
